### PR TITLE
New PrepExternalSchema Implementation

### PR DESCRIPTION
**Pull request by rfm-targa**
### New implementation of PrepExternalSchema

`PrepExternalSchema` moved into a separate module/directory and the adaptation of external schemas no longer uses the [init_schema_4_bbaca](https://github.com/B-UMMI/chewBBACA/blob/master/CHEWBBACA/utils/init_schema_4_bbaca.py) script located in the [utils](https://github.com/B-UMMI/chewBBACA/tree/master/CHEWBBACA/utils) directory.

### Brief description:

- Better handling of input distribution per available core speeds up the process;
- From the alleles that code for the same protein, only one is selected to go through BLASTp, reducing the number of BLASTp searches and matches;
- Representatives are BLASTed against all remaining alleles for a gene/locus, instead of BLASTing representatives against each remaining allele separately. This leads to less BLASTp searches and matches, considerably reducing the run time;
- Invalid alleles and invalid genes listing in output files;
- Output file with total number of alleles, total number of valid alleles and total number of selected representatives, per gene;
- Option to filter sequences based on length and possibility to select genetic code used for translation (by default, genetic code 11 for Bacteria and Archaea).

### TLDR description:

- Better input distribution per available core: input files are distributed into a number of sub-lists equal to the number of available cores. Input distribution is balanced based on the number of sequences per file. This change reduces core idling and avoids the naive input distribution performed by Python;
- Users can provide the BSR value for the adaptation of the external schema (default=0.6);
- The process includes a simple progress bar that updates as each sublist of inputs is processed;
- The process creates a file with the list of invalid alleles according to the following criteria:
    - ambiguous or invalid characters;
    - sequence length is not a multiple of 3;
    - in frame stop codon;
    - invalid start or stop codons.

- The file with the list of invalid alleles includes the reason/reasons that led to considering the listed alleles as invalid;
- A file with the list of genes/loci that did not have any valid alleles is also created;
- Output file with total number of alleles, total number of valid alleles and total number of selected representatives, per gene;
- Option to filter sequences based on length and possibility to select genetic code used for translation (by default, genetic code 11 for Bacteria and Archaea);
- Alleles that code for the same protein are grouped and only one of those alleles is chosen to include in the set of sequences that will go through BLASTp to choose representatives (choosing only one is valid since those alleles code for the same protein and in BLASTp they are equivalent). This serves to reduce the number of BLASTp searches and speed up the process (the alleles grouped by protein are all included in the final schema);
- The first representative that is chosen is the longest, or one of the longest, so that it corresponds to the behavior of the CreateSchema process;
- The first representative is BLASTed against all sequences of that locus in order to determine candidate sequences that might qualify as representatives;
- BLASTp hits are divided into three groups according to their BSR value:
    - High-BSR: are removed from the set of sequences that needs to be BLASTed because the current representative already ensures that they are represented (this set includes every sequence that had a high BSR value with at least one representative);
    - Low-BSR: are kept in the set of sequences that needs to be BLASTed because the current representative is not suited to be their representative (this set includes sequences that only had low BSR hits with representatives);
    - Hotspot-BSR: the hits that have a BSR value between the defined BSR value and the defined BSR value + 0.1 (by default, 0.6 and 0.7). These hits are considered representatives candidates (this set does not include hits with high BSR).
- If there is only one candidate, it is added as a new representative;
- If there are no candidates, the longest sequence of the ones that are not yet represented is added as representative;
- If there is more than one representative, the longest or one of the longest is added as representative;
- The process follows this cycle of representatives being BLASTed against non-represented sequences and addition of one representative until all sequences have a valid representative. For most genes/loci, a single BLAST (or a few) of the representative sequence against all sequences in the gene/locus is enough;
- The adaptation of external schemas is considerably faster and memory usage is low due to the creation of intermediate files for each gene/locus file that is processed and the use of iterators/generators;
- The `PrepExternalSchema` module does not import any functions from other `chewBBACA` modules, including all necessary functions in the main file. This might change in the future as some functions were designed with flexibility in mind and may be included in a single file in order to be imported by different `chewBBACA` processes;
- The new implementation can also be used to add genes/loci to an existing schema and with a few tweaks can be used to add new alleles to genes/loci files in an existing schema.

I appreciate feedback, requests for more information, suggestions to include in new features and any changes that might improve this pull request.

**Tested with:**
Python 3.7.3
BLAST 2.9.0+
BioPython 1.74
Kubuntu 18.10
Intel Core i7-8550U CPU (6 threads)
16GB RAM
500GB SSD

**Testing results:**

| External schema                    | #loci |       #alleles        | time (m/s)  | memory (MB) |
|                      :---:                  | :---:  |           :---:          |      :---:      |        :---:        |
| _Streptococcus agalactiae_<br/>(chewBBACA tutorial)        | 3130 |        17870         |  0m21s      |     34MB        |
| _Clostridioides difficile_<br/>(Ridom cgMLST)              | 2270 |      248179         |    1m22s    |       36MB      |
| _Escherichia coli_<br/>(Ridom cgMLST)                      | 3152 |        529725       |  2m32s      |     37MB        |
| _Salmonella enterica_<br/>(Enterobase cgMLST)               | 3002 |       4907790       |  30m37s    |     132MB       |
| _Salmonella enterica_<br/>(Enterobase wgMLST)               | 21065 |      8264514        |  104m26s    |     544MB       |

After schema adaptation, I ran the AlleleCall process with the adapted schemas, using 2 complete genomes of each schema's species. The AlleleCall process completed successfully in all cases, showing that the schemas were correctly adapted.

**Complete genomes list:**

**_Streptococcus agalactiae_**
[CP019814.1](https://www.ncbi.nlm.nih.gov/assembly/GCA_002881395.1/)
[CP019817.1](https://www.ncbi.nlm.nih.gov/assembly/GCA_002881435.1/)

**_Clostridioides difficile_**
[NZ_CM000441.1](https://www.ncbi.nlm.nih.gov/assembly/GCF_000003215.1/)
[NC_009089.1](https://www.ncbi.nlm.nih.gov/assembly/GCF_000009205.2/)

**_Escherichia coli_**
[NC_000913.3](https://www.ncbi.nlm.nih.gov/assembly/GCF_000005845.2/)
[NC_011750.1](https://www.ncbi.nlm.nih.gov/assembly/GCF_000026345.1/)

**_Salmonella enterica_**
[NC_003197.2](https://www.ncbi.nlm.nih.gov/assembly/GCF_000006945.2/)
[NZ_CP009102.1](https://www.ncbi.nlm.nih.gov/assembly/GCF_000743055.1/)

### CODE REVIEW

**Comment by cimendes** [comment]
Please add discription. Thanks! :skull:

**Comment by cimendes** [comment]
Also, I noticed that the original [init_schema_4_bbaca](https://github.com/B-UMMI/chewBBACA/blob/master/CHEWBBACA/utils/init_schema_4_bbaca.py) has not been removed. Maybe a warning stating that this util has been deprecated would be a nice addition.

**Comment by rfm-targa** [comment]
> Please add description. Thanks! skull

Done :octocat: might add more detail soon.

File: CHEWBBACA/__init__.py [hunk 1]
```diff
@@ -1 +1 @@
-__version__ = "2.0.17.2"
+__version__ = "2.1.0"
```
Comment (cimendes) [inline CHEWBBACA/__init__.py]: :+1:  Good to be in agreement with (semantic versioning guidelines)[https://semver.org/]

File: CHEWBBACA/PrepExternalSchema/PrepExternalSchema.py [hunk 2]
```diff
@@ -0,0 +1,1111 @@
...
+
+    # filter files based on suffix
+    file_suffixes = ['.fasta', '.fna', '.ffn']
+    genes_list = filter_files(genes_list, file_suffixes)
+
+    # filter files based on contents, must be FASTA with sequences
+    genes_list = filter_non_fasta(genes_list)
+
+    # count number of sequences and mean length per gene
+    genes_info = gene_seqs_info(genes_list)
```
Comment (cimendes) [inline CHEWBBACA/PrepExternalSchema/PrepExternalSchema.py]: A minor suggestion of adding multiprocessing here, as well as some feedback for the terminal.

**Comment by cimendes** [comment]
@rfm-targa While running ` python chewBBACA/CHEWBBACA/PrepExternalSchema/PrepExternalSchema.py -i ~/../rmamede/scratch/PrepExternalTest/senterica_enterobase/ -o out/ --cpu 16` the output files generated were named `_invalid_alleles.txt ` and `_invalid_genes.txt`. This is a minor issue, but I think you need a strip somewhere. :P That `-o out/` isn't agreeing much with your code.

**Comment by rfm-targa** [comment]
> Also, I noticed that the original [init_schema_4_bbaca](https://github.com/B-UMMI/chewBBACA/blob/master/CHEWBBACA/utils/init_schema_4_bbaca.py) has not been removed. Maybe a warning stating that this util has been deprecated would be a nice addition.

Done. If the `init_schema_4_bbaca.py` script in the `utils` folder is executed, it will print a warning to the stdout so that users/developers know that the script will be removed from the program in a future version.

**Comment by rfm-targa** [comment]
> @rfm-targa While running ` python chewBBACA/CHEWBBACA/PrepExternalSchema/PrepExternalSchema.py -i ~/../rmamede/scratch/PrepExternalTest/senterica_enterobase/ -o out/ --cpu 16` the output files generated were named `_invalid_alleles.txt ` and `_invalid_genes.txt`. This is a minor issue, but I think you need a strip somewhere. :P That `-o out/` isn't agreeing much with your code.

I think this is solved now! Thanks for telling me.
